File: frontend-web/src/app/services/venta.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class VentaService {
  private http = inject(HttpClient);
  private authService = inject(AuthService);
  private apiUrl = `${environment.apiUrl}/ventas`;

  private get headers(): HttpHeaders {
    return new HttpHeaders({
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.authService.token}`
    });
  }

  checkout(data: {
    metodo_entrega: string;
    direccion_envio?: string;
    sucursal_retiro_id?: number;
  }): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/checkout`, data, { headers: this.headers });
  }

  pagarVenta(ventaId: number, data: { metodo_pago: string; monto: number }): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/pagar/${ventaId}`, data, { headers: this.headers });
  }

  ventaPos(data: {
    sucursal_id: number;
    metodo_pago: string;
    items: { variante_id: number; cantidad: number; precio_unitario: number }[];
    cliente_id?: number;
  }): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/pos`, data, { headers: this.headers });
  }

  getMisCompras(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/mis-compras`, { headers: this.headers });
  }
}

File: frontend_mobile/lib/services/venta_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../constants.dart';

class VentaService {
  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('jwt_token');
  }

  Future<List<dynamic>?> getMisCompras() async {
    try {
      final token = await _getToken();
      if (token == null) return null;
      final response = await http.get(
        Uri.parse('${Constants.apiUrl}/ventas/mis-compras'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) {
      return null;
    }
  }

  Future<Map<String, dynamic>?> checkout(String metodoEntrega, String? direccionEnvio) async {
    try {
      final token = await _getToken();
      final response = await http.post(
        Uri.parse('${Constants.apiUrl}/ventas/checkout'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'metodo_entrega': metodoEntrega,
          if (direccionEnvio != null && direccionEnvio.isNotEmpty) 'direccion_envio': direccionEnvio,
        }),
      );
      if (response.statusCode == 201) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }

  Future<Map<String, dynamic>?> pagarVenta(int ventaId, String metodoPago, double monto) async {
    try {
      final token = await _getToken();
      final response = await http.post(
        Uri.parse('${Constants.apiUrl}/ventas/pagar/$ventaId'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'metodo_pago': metodoPago,
          'monto': monto,
        }),
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }

  Future<Map<String, dynamic>?> ventaPos({
    required int sucursalId,
    required String metodoPago,
    required List<Map<String, dynamic>> items,
    int? clienteId,
  }) async {
    try {
      final token = await _getToken();
      final response = await http.post(
        Uri.parse('${Constants.apiUrl}/ventas/pos'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'sucursal_id': sucursalId,
          'metodo_pago': metodoPago,
          'items': items,
          if (clienteId != null) 'cliente_id': clienteId,
        }),
      );
      if (response.statusCode == 201) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }
}
